src/util/result.h

```diff
@@ -155,25 +194,48 @@ class Result : public detail::ResultBase<T, F>
     template <typename... Args>
     Result(Args&&... args)
     {
-        Construct(std::forward<Args>(args)...);
+        Construct([this](auto&&... x) { this->ConstructValue(std::forward<decltype(x)>(x)...); }, std::forward<Args>(args)...);
     }
     template<typename OT, typename OF>
     Result(Result<OT, OF>&& other) { MoveConstruct(other); }
     Result& operator=(Result&& other) LIFETIMEBOUND
     {
-        if (*this) this->DestroyValue(); else this->m_info.reset();
+        if (*this) this->DestroyValue(); else this->m_info->failure.reset();
         MoveConstruct(other);
         return *this;
     }
 
-    inline friend bilingual_str _ErrorString(const Result& result)
+    void AddError(bilingual_str error)
     {
-        return result ? bilingual_str{} : result.m_info->error;
+        if (!error.empty()) this->Info().errors.emplace_back(std::move(error));
+    }
+
+    void AddWarning(bilingual_str warning)
+    {
+        if (!warning.empty()) this->Info().warnings.emplace_back(std::move(warning));
+    }
+
+    //! Operator moving warning and error messages from this result object to
+    //! another one. Only moves message strings, does not change success or
+    //! failure values of either Result object.
```

Comment:
52a4e50fb4b1171ee0f6814b0a50bc70cdd77134: I don't like that this takes over error strings, but leaves the value and failure untouched. It seems fine to have a result and warnings, but having a result and also an error at the same time seems odd.

Same with operator=. I think this is the first time I've seen that after calling `=`, state is preserved from before `=` was called.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/25665#discussion_r970561659

> [52a4e50](https://github.com/bitcoin/bitcoin/commit/52a4e50fb4b1171ee0f6814b0a50bc70cdd77134): I don't like that this takes over error strings, but leaves the value and failure untouched. It seems fine to have a result and warnings, but having a result and also an error at the same time seems odd.

Agree having both a result value and an error message should be avoided. Also having neither a result value nor an error message should be avoided. But there are tradeoffs around where and how strictly you want to enforce these things. The main thing currently enforcing value/error consistency is having a constructor that sets error messages and failure values at the same time and does not allow setting a result value, and a constructor that only sets a success value and does not allow setting an error message.

But this leaves open the question of what helpers functions like `operator>>` should do when they combine multiple result values that have already been constructed.

The use-case for `operator>>` is when you have an outer function returning `Result<T, F>` calling inner functions returning `Result<T1, F1>`, returning `Result<T2, F2>`, etc. Examples would be [`LoadWalletInternal`](https://github.com/ryanofsky/bitcoin/blob/4e533926fe5549aafb70caf94718447ea497c84c/src/wallet/wallet.cpp#L229), [`CreateWallet`](https://github.com/ryanofsky/bitcoin/blob/4e533926fe5549aafb70caf94718447ea497c84c/src/wallet/wallet.cpp#L298), and [`DoMigration`](https://github.com/ryanofsky/bitcoin/blob/4e533926fe5549aafb70caf94718447ea497c84c/src/wallet/wallet.cpp#L3866) from #25722. The outer function can handle failures from inner functions it calls any way it wants: passing failure values up to its caller, translating failure values, ignoring failures, retrying after failures, falling back to an alternate approaches, etc. It sets success and failure values directly, and it can use `operator>>` to collect error and warning messages separately and pass them along. I don't think `operator>>` should be involved in success and failure value handling. I think it would be bad if `operator>>` discarded error messages, or it it threw runtime exceptions, instead of just passing messages on to ultimately get displayed or logged.

In C++ generally `operator>>` is used for things as varied as bit shifting and stream I/O and can be interpreted as "move data from this place to another place" so I think it reasonable that this `operator>>` just moves error and warning messages from one `Result` object to another, as long as behavior is clearly documented.

> Same with operator=. I think this is the first time I've seen that after calling `=`, state is preserved from before `=` was called.

Yes it is true that assigning to an existing result does not erase warning and errors messages already accumulated in the result. It only sets the success or failure value and appends new warning and error messages to existing ones.

I think this this behavior is safe and useful. Setting a value should not automatically erase warning and error messages that are meant to be displayed or logged. But if this behavior is too surprising for an `operator=` method, we don't actually need to make `Result` assignable, and could rename `operator=()` to `Set()` or `SetValue()`. It looks like even after #25722 there are only 3 `operator=` calls in the codebase outside of tests, so this would be an easy change.

Reply:
> Also having neither a result value nor an error message should be avoided.

Well, this is already impossible for the reasons you gave.

> Agree having both a result value and an error message should be avoided.

Then, why not make it impossible? If there is an outer function returning `Result<T, F>` that wants to combine `Result<T1, F1>` and `Result<T2, F2>`, then it seems better if it explicitly takes (moves) the result T1/F1 out and translates it into T/F. For example, if it passes up a failure value, it seems best to just create a fresh Result (of the outer type) with the failure and return that. (Same if it translates failure values). If it ignores failures, it would be good to translate them to warnings first and not blindly take them over as errors with the `>>` operator. (Same if it retries or falls back).

> I think this this behavior is safe and useful.

Why couldn't the same be achieved by explicitly constructing a new Result with either an error or a value?

Reply:
>> Also having neither a result value nor an error message should be avoided.

> Well, this is already impossible for the reasons you gave.

"Should be avoided" means that users should avoid it, and the implementation makes it easy to avoid by default. The reasons I gave were reasons why a function combining multiple result values needs to either (1) allow result values and error messages to exist at the same time or (2) discard error messages or result values or (3) throw exceptions. And I believe the best choice for `operator>>` is (1), just to be a plain message mover that moves message strings and leaves result values alone. I linked to use-cases showing how this works in practice.

> Agree having both a result value and an error message should be avoided.

> Then, why not make it impossible? If there is an outer function returning `Result<T, F>` that wants to combine `Result<T1, F1>` and `Result<T2, F2>`, then it seems better if it explicitly takes (moves) the result T1/F1 out and translates it into T/F. For example, if it passes up a failure value, it seems best to just create a fresh Result (of the outer type) with the failure and return that.

This is actually what the implementation does. But if the outer function returns success value, and there are error messages, I don't think it is good default behavior to throw away the error messages or raise an exception. I think the best default behavior is to keep error messages so they can be displayed or logged.

> (Same if it translates failure values). If it ignores failures, it would be good to translate them to warnings first and not blindly take them over as errors with the `>>` operator. (Same if it retries or falls back).

So change you are asking for there is basically:

```diff
--- a/src/util/result.h
+++ b/src/util/result.h
@@ -222,7 +222,7 @@ public:
     Result&& operator>>(O&& other LIFETIMEBOUND) &&
     {
         if (this->m_info) {
-            if (!this->m_info->errors.empty()) detail::MoveMessages(this->m_info->errors, other.Info().errors);
+            if (!this->m_info->errors.empty()) detail::MoveMessages(this->m_info->errors, other ? other.Info().warnings : other.Info().errors);
             if (!this->m_info->warnings.empty()) detail::MoveMessages(this->m_info->warnings, other.Info().warnings);
         }
         return std::move(*this);
```

I wouldn't object to it, but it just seems more invasive and doesn't offer practical benefits.

> > I think this this behavior is safe and useful.

> Why couldn't the same be achieved by explicitly constructing a new Result with either an error or a value?

I think

```c++
WarnFn1() >> result;
WarnFn2() >> result;
result = FailFn(...);
return result;
```

or


```c++
WarnFn1() >> result;
WarnFn2() >> result;
result.Set(FailFn(...));
return result;
``` 

is cleaner than

```c++
WarnFn1() >> result;
WarnFn2() >> result;
auto result2 = FailFn(...);
std::move(result) >> result2;
return result2;
```

because it doesn't require introducing multiple result variables. If you are trying to get rid of both `operator=` and `operator>>`, I believe `operator=` or `Set` is also better than:

```c++
auto result2 = FailFn(...);
return result2 ? Result<int, FnError>(std::move(result), result2.value()) : Result<int, FnError>(std::move(result), Error{}, resul2t.GetFailure());
```

I'm happy to rename `operator=` to `Set` if you think `operator=` is misleading. But if you look at the places where these functions are used, it is easier to see why they are useful. Conversely, if you think there is a footgun here, it would be helpful to see an example of the footgun.

Reply:
> But if you look at the places where these functions are used, it is easier to see why they are useful.

I linked to some places where operator<< is used already: [`LoadWalletInternal`](https://github.com/ryanofsky/bitcoin/blob/4e533926fe5549aafb70caf94718447ea497c84c/src/wallet/wallet.cpp#L229), [`CreateWallet`](https://github.com/ryanofsky/bitcoin/blob/4e533926fe5549aafb70caf94718447ea497c84c/src/wallet/wallet.cpp#L298), and [`DoMigration`](https://github.com/ryanofsky/bitcoin/blob/4e533926fe5549aafb70caf94718447ea497c84c/src/wallet/wallet.cpp#L3862) from #25722

For `operator=` (again happy to rename this to `Set`) examples are: [`AddressTableModel::addRow`](https://github.com/ryanofsky/bitcoin/blob/4e533926fe5549aafb70caf94718447ea497c84c/src/qt/addresstablemodel.cpp#L381), [`SQLiteDatabase::Verify`](https://github.com/ryanofsky/bitcoin/blob/4e533926fe5549aafb70caf94718447ea497c84c/src/wallet/sqlite.cpp#L172), [`AvailableCoinsTestingSetup`](https://github.com/ryanofsky/bitcoin/blob/4e533926fe5549aafb70caf94718447ea497c84c/src/wallet/test/availablecoins_tests.cpp#L78), [`FuzzedWallet::GetScriptPubKey`](https://github.com/ryanofsky/bitcoin/blob/4e533926fe5549aafb70caf94718447ea497c84c/src/wallet/test/fuzz/notifications.cpp#L72), [`CWallet::Create`](https://github.com/ryanofsky/bitcoin/blob/4e533926fe5549aafb70caf94718447ea497c84c/src/wallet/wallet.cpp#L2819)

Reply:
Went ahead and renamed `operator=` to `Set` for now. Seems like a good way to avoid some confusion.